### PR TITLE
PARQUET-1970: Make minor releases source compatible

### PR DESCRIPTION
**Pull request by gszadovszky**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: parquet-common/src/main/java/org/apache/parquet/io/OutputFile.java [hunk 1]
```diff
@@ -31,5 +31,7 @@
 
   long defaultBlockSize();
 
-  String getPath();
+  default String getPath() {
+    throw new UnsupportedOperationException();
```
Comment (ggershinsky) [inline parquet-common/src/main/java/org/apache/parquet/io/OutputFile.java:35]: We're consuming this function in one of the encryption options:
https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetWriter.java#L284
If there is no encryption, or encryption is used with a crypto factory that doesn't need the file path - this exception will break the existing callers. A simple way to resolve this is by catching the exception here, and passing a null file path. This allows for these callers to run ok. The crypto factories that do need the file path sometimes (like the one we have in parquet-mr) should check `path==null` in situations where the file path is used, and throw an exception there. What do you think?

Comment (gszadovszky) [reply parquet-common/src/main/java/org/apache/parquet/io/OutputFile.java:35]: It sounds good to me. I'll exchange the default implementation to return `null`.

Comment (ggershinsky) [reply parquet-common/src/main/java/org/apache/parquet/io/OutputFile.java]: Cool, thanks. I'll add then the null checks in the crypto-factory (key management).

Comment (gszadovszky) [reply parquet-common/src/main/java/org/apache/parquet/io/OutputFile.java]: I think, I've added them. Please review the latest commit.

Comment (ggershinsky) [reply parquet-common/src/main/java/org/apache/parquet/io/OutputFile.java]: Oh, this is already covered in the commit. Looks good!

**Comment by gszadovszky** [comment]
@shangxinli, if you have time could you please review this? The release is blocked on this one.

File: parquet-common/src/main/java/org/apache/parquet/io/OutputFile.java (hunk 1 above)
Comment (shangxinli) [reply parquet-common/src/main/java/org/apache/parquet/io/OutputFile.java]: +1 for using null instead of through exception, which is expensive.

**Review by shangxinli — APPROVED** [review]

**Review by Fokko — APPROVED** [review]

**Pull request merged**